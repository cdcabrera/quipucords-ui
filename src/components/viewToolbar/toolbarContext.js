import { storeHooks } from '../../redux';

/**
 * Return filter queries.
 *
 * @param {string} viewId
 * @param {Array} options
 * @alias {object} options
 * @param {Function} options.useSelectors
 * @returns {object}
 */
const useToolbarFieldQueries = (
  viewId,
  options,
  { useSelectors: useAliasSelectors = storeHooks.reactRedux.useSelectors } = {}
) =>
  useAliasSelectors(
    options?.map(({ value }) => ({
      id: value,
      selector: ({ viewOptions }) => viewOptions?.[viewId]?.filterQuery?.[value]
    })),
    {}
  );

/**
 * Return filtered category options, current, and initial value.
 *
 * @param {string} viewId
 * @param {Array} categoryFields
 * @param {object} options
 * @param {Function} options.useSelector
 * @returns {object}
 */
const useSelectCategoryOptions = (
  viewId,
  categoryFields,
  { useSelector: useAliasSelector = storeHooks.reactRedux.useSelector } = {}
) => {
  const updatedCategory = useAliasSelector(({ viewOptions }) => viewOptions?.[viewId]?.currentCategory, null);
  let initialCategory;

  const updatedCategoryFields = categoryFields.map(({ selected, ...option }) => {
    if (updatedCategory === null && selected) {
      initialCategory = option.value;
    }

    return {
      ...option,
      selected: (updatedCategory === null && selected) || updatedCategory === option.value
    };
  });

  return {
    categoryFields: updatedCategoryFields,
    currentCategory: updatedCategory,
    initialCategory
  };
};

/**
 * Clear a specific toolbar category.
 *
 * @returns {Function}
 */
const useToolbarFieldClear = () => field => console.log('CLEAR FIELD >>>>', field);

/**
 * Clear all available toolbar categories.
 *
 * @returns {Function}
 */
const useToolbarFieldClearAll = () => () => console.log('CLEAR FIELDS >>>>');

const context = {
  useSelectCategoryOptions,
  useToolbarFieldClear,
  useToolbarFieldClearAll,
  useToolbarFieldQueries
};

export {
  context as default,
  context,
  useSelectCategoryOptions,
  useToolbarFieldClear,
  useToolbarFieldClearAll,
  useToolbarFieldQueries
};
