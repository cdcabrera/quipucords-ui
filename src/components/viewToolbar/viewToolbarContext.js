import { reduxTypes, storeHooks } from '../../redux';
import { useView } from '../view/viewContext';

/**
 * Clear a specific toolbar category.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useView
 * @returns {Function}
 */

const useToolbarFieldClear = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useView: useAliasView = useView
} = {}) => {
  const { viewId } = useAliasView();
  const dispatch = useAliasDispatch();

  return filter =>
    dispatch([
      {
        type: reduxTypes.view.RESET_PAGE,
        viewId
      },
      {
        type: reduxTypes.view.SET_QUERY,
        viewId,
        filter,
        value: undefined
      }
    ]);
};

/**
 * Clear all available toolbar categories.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useView
 * @returns {Function}
 */
const useToolbarFieldClearAll = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useView: useAliasView = useView
} = {}) => {
  const dispatch = useAliasDispatch();
  const { viewId, config } = useAliasView();
  const options = config?.toolbar?.filterFields;

  return () => {
    const resetFilters = [
      {
        type: reduxTypes.view.RESET_PAGE,
        viewId
      }
    ];

    options.forEach(({ value: filter }) => {
      resetFilters.push({
        type: reduxTypes.view.SET_QUERY,
        viewId,
        filter,
        value: undefined
      });
    });

    dispatch(resetFilters);
  };
};

const context = {
  useToolbarFieldClear,
  useToolbarFieldClearAll
};

export { context as default, context, useToolbarFieldClear, useToolbarFieldClearAll };
