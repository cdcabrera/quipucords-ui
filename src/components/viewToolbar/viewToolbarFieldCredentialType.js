import React from 'react';
import PropTypes from 'prop-types';
import { DropdownSelect } from '../dropdownSelect/dropdownSelect';
import { reduxTypes, storeHooks } from '../../redux';
import { translate } from '../i18n/i18n';

/*
 * Select field options.
 *
 * @type {{title: Function|string, value: string}[]}
 */
const fieldOptions = [
  { title: () => translate('form-dialog.label', { context: ['option', 'network'] }), value: 'network' },
  { title: () => translate('form-dialog.label', { context: ['option', 'satellite'] }), value: 'satellite' },
  { title: () => translate('form-dialog.label', { context: ['option', 'vcenter'] }), value: 'vcenter' }
];

/**
 * On select category for sorting.
 *
 * @param {string} viewType
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnSelect = (viewType, { useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return ({ value = null }) => {
    dispatch([
      {
        type: reduxTypes.viewToolbar.SET_FILTER,
        viewType,
        param: 'cred_type',
        value
      }
    ]);

    /* what it does currently... sets a filter type, and filter value, AND uses "ADD_FILTER" to update the active filters,
        but also concat the heck out of the query string.
    dispatch([
      {
        type: reduxTypes.viewToolbar.SET_FILTER_TYPE,
        viewType,
        filterType: 'cred_type'
      },
      {
        type: reduxTypes.viewToolbar.SET_FILTER_VALUE,
        viewType,
        filterValue: value
      },
      {
        type: reduxTypes.viewToolbar.ADD_FILTER,
        viewType
        filter: { field, value, label: filterText }
      }
    ]);
     */
  };
};

/**
 * Display available credential types.
 *
 * @param {object} props
 * @param {Array} props.options
 * @param {string} props.placeholder
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useSelector
 * @param {string} props.viewType
 * @returns {React.ReactNode}
 */
const ViewToolbarFieldCredentialType = ({
  options,
  placeholder,
  t,
  useOnSelect: useAliasOnSelect,
  useSelector: useAliasSelector,
  viewType
} = {}) => {
  const selectedOption = useAliasSelector(({ viewOptions }) => viewOptions?.[viewType]?.filterQuery?.cred_type);
  const onSelect = useAliasOnSelect(viewType);

  return (
    <DropdownSelect
      ariaLabel={placeholder || t('toolbar.label', { context: ['placeholder', 'filter', 'cred_type'] })}
      placeholder={placeholder || t('toolbar.label', { context: ['placeholder', 'filter', 'cred_type'] })}
      options={options}
      onSelect={onSelect}
      selectedOptions={selectedOption}
    />
  );
};

/**
 * Prop types
 *
 * @type {{useOnSelect: Function, options: Array, placeholder: string}}
 */
ViewToolbarFieldCredentialType.propTypes = {
  placeholder: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
      value: PropTypes.any,
      selected: PropTypes.bool
    })
  ),
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useSelector: PropTypes.func,
  viewType: PropTypes.string
};

/**
 * Default props
 *
 * @type {{useOnSelect: Function, t: translate, options: {title: (Function|string), value: string}[],
 *     placeholder: null}}
 */
ViewToolbarFieldCredentialType.defaultProps = {
  placeholder: null,
  options: fieldOptions,
  t: translate,
  useOnSelect,
  useSelector: storeHooks.reactRedux.useSelector,
  viewType: null
};

export { ViewToolbarFieldCredentialType as default, ViewToolbarFieldCredentialType, fieldOptions, useOnSelect };
