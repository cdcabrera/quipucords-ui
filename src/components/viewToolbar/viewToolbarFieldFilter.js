import React from 'react';
import PropTypes from 'prop-types';
import { InputGroup } from '@patternfly/react-core';
import _debounce from 'lodash/debounce';
import { reduxTypes, storeHooks } from '../../redux';
import { TextInput } from '../form/textInput';
import { translate } from '../i18n/i18n';

/**
 * On submit input, dispatch type.
 *
 * @param {string} viewType
 * @param {string} queryType
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnSubmit = (
  viewType,
  queryType,
  { useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}
) => {
  const dispatch = useAliasDispatch();

  return value =>
    dispatch([
      {
        type: reduxTypes.viewToolbar.SET_FILTER,
        viewType,
        param: queryType,
        value
      }
    ]);
};

/**
 * On clear input, dispatch type.
 *
 * @param {string} viewType
 * @param {string} queryType
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useSelector
 * @returns {Function}
 */
const useOnClear = (
  viewType,
  queryType,
  {
    useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
    useSelector: useAliasSelector = storeHooks.reactRedux.useSelector
  } = {}
) => {
  const currentValue = useAliasSelector(({ viewOptions }) => viewOptions?.[viewType]?.filterQuery?.[queryType]);
  const dispatch = useAliasDispatch();

  return () => {
    if (currentValue === '' || !currentValue) {
      return;
    }

    dispatch([
      {
        type: reduxTypes.viewToolbar.SET_FILTER,
        viewType,
        param: queryType,
        value: ''
      }
    ]);
  };
};

/**
 * Display an input field for filtering results.
 *
 * @fires onKeyUp
 * @param {object} props
 * @param {string} props.queryType
 * @param {Function} props.t
 * @param {number} props.debounceTimer
 * @param {Function} props.useOnClear
 * @param {Function} props.useOnSubmit
 * @param {Function} props.useSelector
 * @param {string} props.viewType
 * @returns {React.ReactNode}
 */
const ViewToolbarFieldFilter = ({
  debounceTimer,
  queryType,
  t,
  useOnClear: useAliasOnClear,
  useOnSubmit: useAliasOnSubmit,
  useSelector: useAliasSelector,
  viewType
}) => {
  const currentValue = useAliasSelector(({ viewOptions }) => viewOptions?.[viewType]?.filterQuery?.[queryType]);
  const onSubmit = useAliasOnSubmit(viewType, queryType);
  const onClear = useAliasOnClear(viewType, queryType);

  /**
   * Set up submit debounce event to allow for bypass.
   */
  const debounced = _debounce(onSubmit, debounceTimer);

  /**
   * On enter submit value, on type submit value, and on esc ignore (clear value at component level).
   *
   * @event onKeyUp
   * @param {object} event
   */
  const onKeyUp = event => {
    switch (event.keyCode) {
      case 13:
        onSubmit(event.value);
        break;
      case 27:
        break;
      default:
        debounced(event.value);
        break;
    }
  };

  return (
    <InputGroup>
      <TextInput
        aria-label={t('toolbar.label', { context: ['placeholder', 'filter', queryType] })}
        className="quipucords-input__search-name"
        iconVariant="search"
        maxLength={255}
        onClear={onClear}
        onKeyUp={onKeyUp}
        value={currentValue}
        placeholder={t('toolbar.label', { context: ['placeholder', 'filter', queryType] })}
        data-test="viewToolbarFieldNameFilter"
      />
    </InputGroup>
  );
};

/**
 * Prop types
 *
 * @type {{useOnSubmit: Function, t: Function, useSelector: Function, debounceTimer: number, viewType: string,
 *     useOnClear: Function, queryType: string}}
 */
ViewToolbarFieldFilter.propTypes = {
  debounceTimer: PropTypes.number,
  queryType: PropTypes.string,
  t: PropTypes.func,
  useOnClear: PropTypes.func,
  useOnSubmit: PropTypes.func,
  useSelector: PropTypes.func,
  viewType: PropTypes.string
};

/**
 * Default props
 *
 * @type {{useOnSubmit: Function, t: translate, useSelector: Function, debounceTimer: number, viewType: null,
 *     useOnClear: Function, queryType: null}}
 */
ViewToolbarFieldFilter.defaultProps = {
  debounceTimer: 800,
  queryType: null,
  t: translate,
  useOnClear,
  useOnSubmit,
  useSelector: storeHooks.reactRedux.useSelector,
  viewType: null
};

export { ViewToolbarFieldFilter as default, ViewToolbarFieldFilter };
