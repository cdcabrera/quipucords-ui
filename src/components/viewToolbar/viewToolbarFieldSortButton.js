import React from 'react';
import { Button, ButtonVariant } from '@patternfly/react-core';
import { SortAmountDownAltIcon, SortAmountUpIcon } from '@patternfly/react-icons';
import PropTypes from 'prop-types';
import { reduxTypes, storeHooks } from '../../redux';
import { API_QUERY_TYPES } from '../../constants/apiConstants';

/**
 * On click sorting.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {string} options.viewId
 * @returns {Function}
 */
const useOnClick = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch, viewId } = {}) => {
  const dispatch = useAliasDispatch();

  return value =>
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_TYPES[(value && API_QUERY_TYPES.ORDERING_DSC) || API_QUERY_TYPES.ORDERING_ASC],
        viewId
      }
    ]);
};

/**
 * Toolbar sort button wrapper.
 *
 * @param {object} props
 * @param {boolean} props.isAscending
 * @param {Function} props.useOnClick
 * @param {Function} props.useSelector
 * @param {string} props.viewId
 * @param {object} props.props
 * @returns {React.ReactNode}
 */
const ViewToolbarFieldSortButton = ({
  isAscending,
  useOnClick: useAliasOnClick,
  useSelector: useAliasSelector,
  viewId,
  ...props
}) => {
  const updatedOrdering = useAliasSelector(
    ({ view }) => view?.query?.[viewId]?.[API_QUERY_TYPES.ORDERING],
    isAscending
  );
  const updatedIsAscending = !/^-/.test(updatedOrdering);
  const onClick = useAliasOnClick({ viewId });

  return (
    <Button
      onClick={() => onClick(!updatedIsAscending)}
      variant={ButtonVariant.plain}
      data-test="toolbarSortButton"
      {...props}
    >
      {(updatedIsAscending && <SortAmountDownAltIcon />) || <SortAmountUpIcon />}
    </Button>
  );
};

/**
 * Prop types
 *
 * @type {{useSelector: Function, viewId: string, useOnClick: Function, isAscending: boolean}}
 */
ViewToolbarFieldSortButton.propTypes = {
  isAscending: PropTypes.bool,
  useOnClick: PropTypes.func,
  useSelector: PropTypes.func,
  viewId: PropTypes.string
};

/**
 * Default props
 *
 * @type {{useSelector: Function, viewId: null, useOnClick: Function, isAscending: boolean}}
 */
ViewToolbarFieldSortButton.defaultProps = {
  isAscending: true,
  useOnClick,
  useSelector: storeHooks.reactRedux.useSelector,
  viewId: null
};

export { ViewToolbarFieldSortButton as default, ViewToolbarFieldSortButton, useOnClick };
