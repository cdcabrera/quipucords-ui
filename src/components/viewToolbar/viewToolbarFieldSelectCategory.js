import React from 'react';
import PropTypes from 'prop-types';
import { useShallowCompareEffect } from 'react-use';
import { FilterIcon } from '@patternfly/react-icons';
import { reduxTypes, storeHooks } from '../../redux';
import { DropdownSelect } from '../dropdownSelect/dropdownSelect';
import { translate } from '../i18n/i18n';
import { useSelectCategoryOptions } from './toolbarContext';

/**
 * On select update category.
 *
 * @param {string} viewId
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnSelect = (viewId, { useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return ({ value = null } = {}) => {
    dispatch([
      {
        type: reduxTypes.viewToolbar.SET_FILTER_TYPE,
        viewId,
        // currentFilter: value
        currentCategory: value
      }
    ]);
  };
};

/**
 * Select available filter categories.
 *
 * @param {object} props
 * @param {Array} props.options
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useSelectCategoryOptions
 * @param {string} props.viewId
 * @returns {React.ReactNode}
 */
const ViewToolbarFieldSelectCategory = ({
  options,
  t,
  useOnSelect: useAliasOnSelect,
  // useSelector: useAliasSelector,
  useSelectCategoryOptions: useAliasSelectCategoryOptions,
  viewId
  // t,
  // useOnSelect: useAliasOnSelect,
  // useSelectCategoryOptions: useAliasSelectCategoryOptions
}) => {
  // const selectedOption = useAliasSelector(({ viewOptions }) => viewOptions?.[viewId]?.currentCategory, undefined);
  const { categoryFields, initialCategory } = useAliasSelectCategoryOptions(viewId, options);
  const onSelect = useAliasOnSelect(viewId);

  useShallowCompareEffect(() => {
    if (initialCategory) {
      onSelect({ value: initialCategory });
    }
  }, [initialCategory, onSelect]);

  return (
    <DropdownSelect
      ariaLabel={t('toolbar.label', { context: ['placeholder', 'filter'] })}
      placeholder={t('toolbar.label', { context: ['placeholder', 'filter'] })}
      options={categoryFields}
      onSelect={onSelect}
      // selectedOptions={currentCategory}
      toggleIcon={<FilterIcon />}
    />
  );
};

/**
 * Prop types.
 *
 * @type {{useOnSelect: Function, t: Function, useSelectCategoryOptions: Function}}
 */
ViewToolbarFieldSelectCategory.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
      value: PropTypes.any,
      selected: PropTypes.bool
    })
  ),
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useSelectCategoryOptions: PropTypes.func,
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{useOnSelect: Function, t: Function, useSelectCategoryOptions: Function}}
 */
ViewToolbarFieldSelectCategory.defaultProps = {
  options: [],
  t: translate,
  useOnSelect,
  useSelectCategoryOptions,
  viewId: null
};

export { ViewToolbarFieldSelectCategory as default, ViewToolbarFieldSelectCategory, useOnSelect };
