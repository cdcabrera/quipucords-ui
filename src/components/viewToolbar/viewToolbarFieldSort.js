import React from 'react';
import PropTypes from 'prop-types';
import { reduxTypes, storeHooks } from '../../redux';
import { ViewToolbarFieldSortButton } from './viewToolbarFieldSortButton';
import { Tooltip } from '../tooltip/tooltip';
import { DropdownSelect } from '../dropdownSelect/dropdownSelect';
import { translate } from '../i18n/i18n';
import { API_QUERY_TYPES } from '../../constants/apiConstants';
import { useQuery, useView } from '../view/viewContext';

/**
 * On select category for sorting.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useView
 * @returns {Function}
 */
const useOnSelect = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useView: useAliasView = useView
} = {}) => {
  const { viewId } = useAliasView();
  const dispatch = useAliasDispatch();

  return ({ value }) =>
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY,
        viewId,
        filter: API_QUERY_TYPES.ORDERING,
        value
      }
    ]);
};

/**
 * Toolbar sort button wrapper.
 *
 * @param {object} props
 * @param {Array} props.options
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useQuery
 * @returns {React.ReactNode}
 */
const ViewToolbarFieldSort = ({ options, t, useOnSelect: useAliasOnSelect, useQuery: useAliasQuery }) => {
  const { [API_QUERY_TYPES.ORDERING]: selectedOption } = useAliasQuery();
  const onSelect = useAliasOnSelect();

  return (
    <React.Fragment>
      <DropdownSelect
        options={options}
        onSelect={onSelect}
        selectedOptions={selectedOption}
        data-test="toolbarSortType"
      />
      <Tooltip placement="right" content={t('toolbar.label', { context: ['option', 'sort', selectedOption] })}>
        <ViewToolbarFieldSortButton />
      </Tooltip>
    </React.Fragment>
  );
};

/**
 * Prop types
 *
 * @type {{useQuery: Function, t: Function, useSelector: Function, options:Array}}
 */
ViewToolbarFieldSort.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
      value: PropTypes.any,
      selected: PropTypes.bool
    })
  ),
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useQuery: PropTypes.func
};

/**
 * Default props
 *
 * @type {{useOnSelect: Function, t: Function, useQuery: Function, options: *[]}}
 */
ViewToolbarFieldSort.defaultProps = {
  options: [],
  t: translate,
  useOnSelect,
  useQuery
};

export { ViewToolbarFieldSort as default, ViewToolbarFieldSort, useOnSelect };
