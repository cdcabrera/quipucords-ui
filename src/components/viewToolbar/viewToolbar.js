import React from 'react';
import PropTypes from 'prop-types';
import {
  Divider,
  Toolbar,
  ToolbarContent,
  ToolbarFilter,
  ToolbarGroup,
  ToolbarItem,
  ToolbarItemVariant,
  ToolbarToggleGroup
} from '@patternfly/react-core';
import { FilterIcon } from '@patternfly/react-icons';
import { storeHooks } from '../../redux';
import { useToolbarFieldClear, useToolbarFieldClearAll } from './viewToolbarContext';
import { useOnRefresh, useView } from '../view/viewContext';
import { ViewToolbarSelectCategory } from './viewToolbarSelectCategory';
import { ViewToolbarFieldSort } from './viewToolbarFieldSort';
import { RefreshTimeButton } from '../refreshTimeButton/refreshTimeButton';
import { translate } from '../i18n/i18n';

const ViewToolbar = ({
  lastRefresh,
  secondaryFields,
  t,
  useOnRefresh: useAliasOnRefresh,
  useSelector: useAliasSelector,
  useToolbarFieldClear: useAliasToolbarFieldClear,
  useToolbarFieldClearAll: useAliasToolbarFieldClearAll,
  useView: useAliasView
}) => {
  const { config, query, viewId } = useAliasView();
  const updatedCategoryFields = config?.toolbar?.filterFields || [];

  const currentCategory = useAliasSelector(({ view }) => view?.filters?.[viewId]?.currentFilterCategory);
  const onRefresh = useAliasOnRefresh();
  const clearField = useAliasToolbarFieldClear();
  const clearAllFields = useAliasToolbarFieldClearAll();

  /**
   * Clear a specific filter
   *
   * @event onClearFilter
   * @param {object} params
   * @param {*} params.value
   * @returns {void}
   */
  const onClearFilter = ({ value }) => clearField(value);

  /**
   * Clear all active filters.
   *
   * @event onClearAll
   * @returns {void}
   */
  const onClearAll = () => clearAllFields();

  /**
   * Set selected options for chip display.
   *
   * @param {*|string} value
   * @returns {Array}
   */
  const setSelectedOptions = value => {
    const categoryValue = query?.[value];
    return (categoryValue && [t('toolbar.label', { context: ['chip', categoryValue] })]) || [];
  };

  return (
    <React.Fragment>
      <Toolbar
        className="quipucords-toolbar"
        collapseListedFiltersBreakpoint="lg"
        clearAllFilters={onClearAll}
        clearFiltersButtonText={t('toolbar.label', { context: 'clear-filters' })}
      >
        <ToolbarContent>
          <ToolbarToggleGroup toggleIcon={<FilterIcon />} breakpoint="lg">
            <ToolbarGroup variant="filter-group">
              {updatedCategoryFields.length > 1 && (
                <ToolbarItem>
                  <ViewToolbarSelectCategory />
                </ToolbarItem>
              )}
              {updatedCategoryFields.map(({ title, value, component: OptionComponent }) => {
                const chipProps = { categoryName: (typeof title === 'function' && title()) || title };
                chipProps.chips = setSelectedOptions(value);
                chipProps.deleteChip = () => onClearFilter({ value });

                return (
                  <ToolbarFilter
                    key={value}
                    showToolbarItem={currentCategory === value || updatedCategoryFields.length === 1}
                    {...chipProps}
                  >
                    <OptionComponent />
                  </ToolbarFilter>
                );
              })}
            </ToolbarGroup>
          </ToolbarToggleGroup>
          <ToolbarItem key="groupSeparator" variant={ToolbarItemVariant.separator} />
          <ToolbarItem key="sortFields" spacer={{ default: 'spacerSm' }}>
            <ViewToolbarFieldSort />
          </ToolbarItem>
          <ToolbarItem key="sortSeparator" variant={ToolbarItemVariant.separator} />
          <ToolbarItem key="lastRefresh">
            <RefreshTimeButton onRefresh={onRefresh} lastRefresh={lastRefresh} />
          </ToolbarItem>
          <ToolbarGroup
            key="secondaryFields"
            align={{ lg: 'alignRight', md: 'alignLeft' }}
            spacer={{ default: 'spacerSm' }}
          >
            {secondaryFields.map(field => (
              <ToolbarItem key={field.key}>{field}</ToolbarItem>
            ))}
          </ToolbarGroup>
        </ToolbarContent>
      </Toolbar>
      <Divider />
    </React.Fragment>
  );
};

ViewToolbar.propTypes = {
  secondaryFields: PropTypes.array,
  lastRefresh: PropTypes.number,
  t: PropTypes.func,
  useOnRefresh: PropTypes.func,
  useSelector: PropTypes.func,
  useToolbarFieldClear: PropTypes.func,
  useToolbarFieldClearAll: PropTypes.func,
  useView: PropTypes.func
};

ViewToolbar.defaultProps = {
  secondaryFields: [],
  lastRefresh: 0,
  t: translate,
  useSelector: storeHooks.reactRedux.useSelector,
  useOnRefresh,
  useToolbarFieldClear,
  useToolbarFieldClearAll,
  useView
};

export { ViewToolbar as default, ViewToolbar };
