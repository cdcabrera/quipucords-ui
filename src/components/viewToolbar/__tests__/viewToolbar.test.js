import React from 'react';
import { ToolbarFilter } from '@patternfly/react-core';
import { ViewToolbar } from '../viewToolbar';
import { ViewToolbarTextInput } from '../viewToolbarTextInput';
import { API_QUERY_TYPES } from '../../../constants/apiConstants';
import { CONFIG as sourcesConfig } from '../../sources/sources';

describe('ViewToolbar Component', () => {
  it('should render a basic component', async () => {
    const props = {
      useView: () => ({ viewId: sourcesConfig.viewId, config: { toolbar: sourcesConfig.toolbar } })
    };
    const component = await shallowHookComponent(<ViewToolbar {...props} />);

    expect(component).toMatchSnapshot('basic');
  });

  it('should hide categories when a single filter is available', async () => {
    const props = {
      useView: () => ({
        viewId: sourcesConfig.viewId,
        config: {
          toolbar: { filterFields: [sourcesConfig.toolbar.filterFields[0]] }
        }
      })
    };
    const component = await mountHookComponent(<ViewToolbar {...props} />);

    expect(component.find(ViewToolbarTextInput).first()).toMatchSnapshot('single filter');
  });

  it('should handle updating toolbar chips', async () => {
    const props = {
      useView: () => ({
        viewId: sourcesConfig.viewId,
        query: { [API_QUERY_TYPES.SEARCH_NAME]: 'lorem ipsum' },
        config: {
          toolbar: { filterFields: [sourcesConfig.toolbar.filterFields[0]] }
        }
      })
    };
    const component = await shallowHookComponent(<ViewToolbar {...props} />);

    expect(component.find(ToolbarFilter).props()).toMatchSnapshot('chips');
  });

  it('should handle displaying secondary components, fields', async () => {
    const props = {
      useView: () => ({
        viewId: sourcesConfig.viewId,
        config: {
          toolbar: { filterFields: [sourcesConfig.toolbar.filterFields[0]] }
        }
      }),
      secondaryFields: <React.Fragment>dolor sit</React.Fragment>
    };
    const component = await shallowHookComponent(<ViewToolbar {...props} />);

    expect(component).toMatchSnapshot('secondary');
  });
});
