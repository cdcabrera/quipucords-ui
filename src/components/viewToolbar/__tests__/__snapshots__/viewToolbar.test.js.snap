// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewToolbar Component should handle displaying secondary components, fields: secondary 1`] = `
<React.Fragment>
  <Toolbar
    className="quipucords-toolbar"
    clearAllFilters={[Function]}
    clearFiltersButtonText="t(toolbar.label, {"context":"clear-filters"})"
    collapseListedFiltersBreakpoint="lg"
  >
    <ToolbarContent
      isExpanded={false}
      showClearFiltersButton={false}
    >
      <ToolbarToggleGroup
        breakpoint="lg"
        toggleIcon={<FilterIcon />}
      >
        <ForwardRef
          variant="filter-group"
        >
          <ToolbarFilter
            categoryName="t(toolbar.label_option, {"context":"search_by_name"})"
            chips={[]}
            deleteChip={[Function]}
            showToolbarItem={true}
          >
            <SearchName />
          </ToolbarFilter>
        </ForwardRef>
      </ToolbarToggleGroup>
      <ToolbarItem
        variant="separator"
      />
      <ToolbarItem
        spacer={
          {
            "default": "spacerSm",
          }
        }
      >
        <ViewToolbarFieldSort
          t={[Function]}
          useOnSelect={[Function]}
          useView={[Function]}
        />
      </ToolbarItem>
      <ToolbarItem
        variant="separator"
      />
      <ToolbarItem>
        <RefreshTimeButton
          lastRefresh={0}
          onRefresh={[Function]}
          t={[Function]}
        />
      </ToolbarItem>
      <ForwardRef
        align={
          {
            "lg": "alignRight",
            "md": "alignLeft",
          }
        }
        spacer={
          {
            "default": "spacerSm",
          }
        }
      >
        <ToolbarItem>
          <React.Fragment>
            dolor sit
          </React.Fragment>
        </ToolbarItem>
      </ForwardRef>
    </ToolbarContent>
  </Toolbar>
  <Divider />
</React.Fragment>
`;

exports[`ViewToolbar Component should handle updating toolbar chips: chips 1`] = `NodeList []`;

exports[`ViewToolbar Component should hide categories when a single filter is available: single filter 1`] = `
NodeList [
  <input
    aria-invalid="false"
    aria-label="t(toolbar.label_placeholder_filter, {"context":"search_by_name"})"
    data-ouia-component-id="toolbar_search_name"
    data-ouia-component-type="PF5/TextInput"
    data-ouia-safe="true"
    id="generatedid-"
    maxlength="255"
    name="generatedid-"
    placeholder="t(toolbar.label_placeholder_filter, {"context":"search_by_name"})"
    type="text"
    value=""
  />,
]
`;

exports[`ViewToolbar Component should render a basic component: basic 1`] = `
<React.Fragment>
  <Toolbar
    className="quipucords-toolbar"
    clearAllFilters={[Function]}
    clearFiltersButtonText="t(toolbar.label, {"context":"clear-filters"})"
    collapseListedFiltersBreakpoint="lg"
  >
    <ToolbarContent
      isExpanded={false}
      showClearFiltersButton={false}
    >
      <ToolbarToggleGroup
        breakpoint="lg"
        toggleIcon={<FilterIcon />}
      >
        <ForwardRef
          variant="filter-group"
        >
          <ToolbarItem>
            <ViewToolbarSelectCategory
              t={[Function]}
              useOnSelect={[Function]}
              useSelector={[Function]}
              useView={[Function]}
            />
          </ToolbarItem>
          <ToolbarFilter
            categoryName="t(toolbar.label_option, {"context":"search_by_name"})"
            chips={[]}
            deleteChip={[Function]}
            showToolbarItem={false}
          >
            <SearchName />
          </ToolbarFilter>
          <ToolbarFilter
            categoryName="t(toolbar.label_option, {"context":"search_credentials_by_name"})"
            chips={[]}
            deleteChip={[Function]}
            showToolbarItem={false}
          >
            <SearchCredentialsName />
          </ToolbarFilter>
          <ToolbarFilter
            categoryName="t(toolbar.label_option, {"context":"source_type"})"
            chips={[]}
            deleteChip={[Function]}
            showToolbarItem={false}
          >
            <SourceType />
          </ToolbarFilter>
        </ForwardRef>
      </ToolbarToggleGroup>
      <ToolbarItem
        variant="separator"
      />
      <ToolbarItem
        spacer={
          {
            "default": "spacerSm",
          }
        }
      >
        <ViewToolbarFieldSort
          t={[Function]}
          useOnSelect={[Function]}
          useView={[Function]}
        />
      </ToolbarItem>
      <ToolbarItem
        variant="separator"
      />
      <ToolbarItem>
        <RefreshTimeButton
          lastRefresh={0}
          onRefresh={[Function]}
          t={[Function]}
        />
      </ToolbarItem>
      <ForwardRef
        align={
          {
            "lg": "alignRight",
            "md": "alignLeft",
          }
        }
        spacer={
          {
            "default": "spacerSm",
          }
        }
      />
    </ToolbarContent>
  </Toolbar>
  <Divider />
</React.Fragment>
`;
